calico_cni_test.go

```diff
@@ -297,4 +299,155 @@ var _ = Describe("CalicoCni", func() {
 			})
 		})
 	})
+
+	Describe("ADD a continaer with ContainerID and DEL it with the same ContainerID", func() {
+		Context("Use the same CNI_ContainerID to ADD and DEL the container", func() {
+			netconf := fmt.Sprintf(`
+            {
+              "cniVersion": "%s",
+              "name": "net1",
+              "type": "calico",
+              "etcd_endpoints": "http://%s:2379",
+              "ipam": {
+                "type": "host-local",
+                "subnet": "10.0.0.0/8"
+              }
+            }`, cniVersion, os.Getenv("ETCD_IP"))
+
+			It("should succesfully ADD and DEL the container with CNI_ContainerID", func() {
+				cniContainerID := "container-id-001"
+
+				// ADD the container with passing a CNI_ContainerID.
+				workloadID, netnspath, session, _, _, _, err := CreateContainerWithId(netconf, "", "", cniContainerID)
+				Expect(err).ShouldNot(HaveOccurred())
+				Eventually(session).Should(gexec.Exit())
+
+				result, err := GetResultForCurrent(session, cniVersion)
+				if err != nil {
+					log.Fatalf("Error getting result from the session: %v\n", err)
+				}
+
+				log.Printf("Unmarshaled result: %v\n", result)
+
+				// The endpoint is created in the backend datastore.
+				endpoints, err := calicoClient.WorkloadEndpoints().List(api.WorkloadEndpointMetadata{})
+				Expect(err).ShouldNot(HaveOccurred())
+				Expect(endpoints.Items).Should(HaveLen(1))
+				Expect(endpoints.Items[0].Metadata).Should(Equal(api.WorkloadEndpointMetadata{
+					Node:             hostname,
+					Name:             "eth0",
+					Workload:         workloadID,
+					ActiveInstanceID: cniContainerID,
+					Orchestrator:     "cni",
+				}))
+
+				// Delete the container with the same CNI_ContainerID.
+				exitCode, err := DeleteContainerWithId(netconf, netnspath, "", cniContainerID)
+				Expect(err).ShouldNot(HaveOccurred())
+				Expect(exitCode).Should(Equal(0))
+
+				// The endpoint should not exist in the backend datastore.
+				endpoints, err = calicoClient.WorkloadEndpoints().List(api.WorkloadEndpointMetadata{})
+				Expect(err).ShouldNot(HaveOccurred())
+				Expect(endpoints.Items).Should(HaveLen(0))
+			})
+		})
+	})
+
+	Describe("ADD a continaer with a ContainerID and DEL it with the same ContainerID then ADD a new container with a different ContainerID, and send a DEL for the old ContainerID", func() {
```

Comment:
right now it's just k8s being too "aggressive" in sending DELs, we could have other non-k8s doing the same and since both of these take different code paths (for ADD) I wanted to make sure we get the same behavior. In fact, I found an issue with this test, as the result I had L181 in calico.go :)